iio/iio.c

```diff
@@ -0,0 +1,838 @@
+		if (!iio_interfaces->interfaces)
+			return -ENOMEM;
+	}
+	iio_interface = calloc(1, sizeof(struct iio_interface));
+	if (!iio_interface)
+		return -ENOMEM;
+
+	iio_interface->dev_instance = init_par->dev_instance;
+	iio_interface->name = init_par->dev_name;
+	iio_interface->iio = init_par->iio_device;
+	iio_interface->get_xml = init_par->get_xml;
+	iio_interface->transfer_dev_to_mem = init_par->transfer_dev_to_mem;
+	iio_interface->transfer_mem_to_dev = init_par->transfer_mem_to_dev;
+	iio_interface->read_data = init_par->read_data;
+	iio_interface->write_data = init_par->write_data;
+
+	iio_interfaces->interfaces[iio_interfaces->num_interfaces - 1] = iio_interface;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Unregister interface.
+ * @param *device_name - String containing device name.
+ * @return SUCCESS in case of success or negative value otherwise.
+ */
+ssize_t iio_unregister(const char *device_name)
+{
+	struct iio_interface *iio_interface = iio_get_interface(device_name,
+					      iio_interfaces);
+	struct iio_interfaces *interfaces;
+	int16_t i, deleted = 0;
+
+	if (!iio_interface)
+		return FAILURE;
+
+	interfaces = calloc(1, sizeof(struct iio_interfaces));
+	if (!interfaces)
+		return FAILURE;
+
+	interfaces->interfaces = calloc(iio_interfaces->num_interfaces - 1,
+					sizeof(struct iio_interface*));
+	if (!interfaces->interfaces) {
+		free(interfaces);
+		return FAILURE;
+	}
+
+	for(i = 0; i < iio_interfaces->num_interfaces; i++) {
+		if (!strcmp(device_name, iio_interfaces->interfaces[i]->name)) {
+			free(iio_interfaces->interfaces[i]);
+			deleted = 1;
+			continue;
+		}
+		interfaces->interfaces[i - deleted] = iio_interfaces->interfaces[i];
+	}
+
+	interfaces->num_interfaces = iio_interfaces->num_interfaces - 1;
+	free(iio_interfaces);
+	iio_interfaces = interfaces;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set communication ops and read/write ops that will be called from
+ * "libtinyiiod".
+ * @param **iiod - Structure containing new tinyiiod instance.
+ * @param iio_server_ops - Structure containing read/write ops (Ex: read/write to UART).
+ * @return SUCCESS in case of success or negative value otherwise.
+ */
+ssize_t iio_init(struct tinyiiod **iiod, struct iio_server_ops *iio_server_ops)
+{
+	struct tinyiiod_ops *ops = calloc(1, sizeof(struct tinyiiod_ops));
+
+	if (!ops)
+		return FAILURE;
+
+	/* device operations */
+	ops->read_attr = iio_read_attr;
+	ops->write_attr = iio_write_attr;
+	ops->ch_read_attr = iio_ch_read_attr;
+	ops->ch_write_attr = iio_ch_write_attr;
+	ops->transfer_dev_to_mem = iio_transfer_dev_to_mem;
+	ops->read_data = iio_read_dev;
+	ops->transfer_mem_to_dev = iio_transfer_mem_to_dev;
+	ops->write_data = iio_write_dev;
+
+	ops->open = iio_open_dev;
+	ops->close = iio_close_dev;
+	ops->get_mask = iio_get_mask;
+
+	ops->read = iio_server_ops->read;
+	ops->write = iio_server_ops->write;
+	ops->get_xml = iio_get_xml;
+
+	*iiod = tinyiiod_create(ops);
+	if (!(*iiod))
+		goto error_free_ops;
```

Comment:
`goto` is very confusing and this function could end wth a simple `if/else`

Reply:
Fixed